### DISCUSSION TITLE
Why isn't this project properly documented?

### DISCUSSION THREAD

**Discussion by paulrmolina** [project64/project64#2403] [Ideas]
I've been going into several header files and there aren't any comments whatsoever on what any of the functions actually does.

The code itself is good and well-documented through its source code, albeit there's a serious lack of commenting in the source itself.

If this is open-source, at least give us the breadcrumbs so we can make improvements!

**Comment by project64** [comment]
if something is confusing or not obvious I will add a comment, but usually when I do look at other projects I find that comments are just documenting something that is super obvious looking at the source code, or it is there cause the code is horribly written. 

I try to make the code clear and read like pseucode so you can easily understand it.

Where do you think it actually needs more documentation, beyond other projects have comments so this should have them too?

**Reply by project64** [reply]
@paulrmolina I got a notification of a comment, but it is not here so I assume you deleted it?

> A professional programmer will comment all of their code, parameters, function... etc.
this is a bit of an overgeneralization, the extent and nature of commenting can vary greatly based on the context.

If we are talking about a public API, like the plugin api then detailed comments explaining each function and parameter is very important. Those do have a lot of comments around them.

>You have a lot of technical debt.
If you mean I just do have many comments as technical debt, then you have to think about why is a meaningful comments that explain the why and how of complex or non-obvious parts of the code, rather than commenting every single line or function. Redundant comments can clutter the code and reduce readability.

Comments need to be maintained along with the code. Outdated or incorrect comments can be more harmful than no comments at all, as they can lead to misunderstandings. So having a lot of comments adds to the work amount.

the code should be self-documenting, reading it should easily explain the code, not using magic numbers (having them as as enum so you can see what it does), good naming of variables and function names so you can understand the code.

An emulator is complicated code, so you need a certain amount of understanding to even be playing with the code.
